docs/ts/primitives/validation.mdx

```diff
@@ -0,0 +1,468 @@
+
+### Reusing the request type as the response type
+It is often the case that you want to return the same data type that you received in the request. In this case, you can reuse the request type as the response type:
+
+```typescript
+import { api, Header, Query } from "encore.dev/api";
+
+interface Data {
+  header: Header<"X-Header">; // this field will be read from the http header
+  query: Query<string>; // this will be parsed from the '?query=...' parameter in the request url
+  body: string; // this will be sent as part of the JSON body
+}
+// A simple API endpoint that echoes the data back.
+export const echo = api(
+  { method: "POST", path: "/echo" },
+  async (params: Data): Promise<Data> => {
+    return params; // echo the data back
+  },
+);
+```
+
+This API endpoint expects incoming requests to look like this:
+```output
+POST /echo?query=hello HTTP/1.1
+Content-Type: application/json
+X-Header: this is a header
+{
+   "body": "a body",
+}
+```
+
+For HTTP responses the `Query<string>` type is considered to be part of the JSON response body, since query strings only make sense for incoming requests.
+Responses returned from this endpoint will be serialized as a HTTP response to looks like this:
+
+```output
+HTTP/1.1 200 OK
+Content-Type: application/json
+X-Header: this is a header
+{
+   "query": "hello",
+   "body": "a body",
+}
+```
+
+## Under the hood
+
+Encore.ts parses your source code to understand the request and response schema that each API endpoint expects, including things like HTTP headers, query parameters, and so on. The schemas are then processed, optimized, and stored as a Protobuf file.
+
+When the Encore Rust Runtime starts up, it reads this Protobuf file and pre-computes a request decoder and response encoder, optimized for each API endpoint, using the exact type definition each API endpoint expects. In fact, Encore.ts even handles request validation directly in Rust, ensuring invalid requests never have to even touch the JS layer, mitigating many denial of service attacks.
+
+Encore’s understanding of the request schema also proves beneficial from a performance perspective. JavaScript runtimes like Deno and Bun use a similar architecture to that of Encore's Rust-based runtime (in fact, Deno also uses Rust+Tokio+Hyper), but lack Encore’s understanding of the request schema. As a result, they need to hand over the un-processed HTTP requests to the single-threaded JavaScript engine for execution.
+
+Encore.ts, on the other hand, handles much more of the request processing inside Rust, and only hands over the decoded request objects. By handling much more of the request life-cycle in multi-threaded Rust, the JavaScript event-loop is freed up to focus on executing application business logic instead of parsing HTTP requests, yielding an even greater performance boost.
```

Comment:
```suggestion
When the Encore.ts Rust runtime starts up, it reads the Protobuf file and pre-computes a request decoder and response encoder, optimized for each API endpoint, using the exact type definition each API endpoint expects. In fact, Encore.ts even handles request validation directly in Rust, ensuring invalid requests never have to even touch the JavaScript layer, mitigating many denial of service attacks.

Encore’s understanding of the request schema also improves performance. JavaScript runtimes like Deno and Bun use a similar architecture (in fact, Deno also uses Rust+Tokio+Hyper), but lack Encore’s understanding of the request schema. As a result, they need to hand over the un-processed HTTP requests to the single-threaded JavaScript engine for execution.

Encore.ts, on the other hand, handles much more of the request processing inside Rust, and only hands over the decoded request objects. By handling much more of the request life-cycle in multi-threaded Rust, the JavaScript event-loop is freed up to focus on executing application business logic instead of parsing HTTP requests, resulting in a significant performance improvement.
```